tests/importer.rs

```diff
@@ -0,0 +1,252 @@
+
+    assert_eq!(importer.path_checksums, HashMap::new());
+    assert_eq!(importer.checksum_paths, HashMap::new());
+    assert_eq!(importer.loaded_models, HashMap::new());
+}
+
+#[test]
+fn test_returns_error_when_importing_invalid_obj_file() {
+    let mut importer = Importer::new();
+    let path = "tests/fixtures/invalid.obj".to_string();
+
+    let error = importer
+        .import_obj(&path)
+        .expect_err("Error should be thrown");
+
+    assert_eq!(error, ImporterError::InvalidStructure);
+}
+
+#[test]
+fn test_does_not_cache_nonexistent_file() {
+    let mut importer = Importer::new();
+    let path = "tests/fixtures/wrong_path.obj".to_string();
+
+    importer
+        .import_obj(&path)
+        .expect_err("Error should be thrown");
+
+    assert_eq!(importer.path_checksums, HashMap::new());
+    assert_eq!(importer.checksum_paths, HashMap::new());
+    assert_eq!(importer.loaded_models, HashMap::new());
+}
+
+#[test]
+fn test_returns_error_when_importing_nonexistent_file() {
+    let mut importer = Importer::new();
+    let path = "tests/fixtures/wrong_path.obj".to_string();
+
+    let error = importer
+        .import_obj(&path)
+        .expect_err("Error should be thrown");
+
+    assert_eq!(error, ImporterError::FileNotFound);
+}
+
+#[test]
+fn test_does_not_cache_the_same_unchanged_obj_file_twice() {
+    let mut importer = Importer::new();
+    let path = "tests/fixtures/valid.obj".to_string();
+    let file_contents = fs::read_to_string(&path).expect("File should be read to string");
+    let checksum = importer::calculate_checksum(&file_contents);
+
+    importer
+        .import_obj(&path)
+        .expect("Valid object should be loaded");
+    importer
+        .import_obj(&path)
+        .expect("Valid object should be loaded");
+
+    let mut expected_path_checksums = HashMap::new();
+    expected_path_checksums.insert(path.clone(), checksum);
+
+    let mut expected_checksum_paths = HashMap::new();
+    expected_checksum_paths.insert(checksum, vec![path.clone()]);
+
+    let loaded_models_paths: HashSet<_> = importer.loaded_models.keys().collect();
+    let mut expected_loaded_models_paths: HashSet<&String> = HashSet::new();
+    expected_loaded_models_paths.insert(&path);
+
+    assert_eq!(importer.path_checksums, expected_path_checksums);
+    assert_eq!(importer.checksum_paths, expected_checksum_paths);
+    assert_eq!(loaded_models_paths, expected_loaded_models_paths);
+}
+
+#[test]
+fn test_returns_correct_models_when_importing_the_same_unchanged_file() {
+    let mut importer = Importer::new();
+    let path = "tests/fixtures/valid.obj".to_string();
+
+    let models_1 = importer
+        .import_obj(&path)
+        .expect("Valid object should be loaded");
+    let models_2 = importer
+        .import_obj(&path)
+        .expect("Valid object should be loaded");
+
+    assert_eq!(models_1, models_2);
+}
+
+#[test]
+fn test_adds_two_different_files_with_the_same_contents() {

```

Comment:
This test actually tests performance, not correctness. Not sure If I would keep it.

Reply:
Name of the function is actually wrong. `adds` -> `caches`.

And since multiple files can have the same checksum, it makes sense to test whether it was properly cached or not.

Reply:
I was basing the comment on the implementation, not the name, but you are right with the name change 👍 

I initially suggested to just remove the test case because I didn't want to over-complicate the API just because a test needs some data, but if you really want to keep the test case, this could work:

We could modify the `has_cached(&str)` I mentioned earlier to also return the number of paths pointing to the model and assert it increments when you load the second model.